src/SAWScript/Builtins.hs

```diff
@@ -1021,17 +1021,23 @@ beta_reduce_term (TypedTerm schema t) = do
   t' <- io $ betaNormalize sc t
   return (TypedTerm schema t')
 
-addsimp :: Theorem -> Simpset -> Simpset
-addsimp (Theorem (Prop t) _stats) ss = addRule (ruleOfProp t) ss
-
-addsimp' :: Term -> Simpset -> Simpset
-addsimp' t ss = addRule (ruleOfProp t) ss
-
-addsimps :: [Theorem] -> Simpset -> Simpset
-addsimps thms ss = foldr addsimp ss thms
-
-addsimps' :: [Term] -> Simpset -> Simpset
-addsimps' ts ss = foldr (\t -> addRule (ruleOfProp t)) ss ts
+addsimp :: Theorem -> Simpset -> TopLevel Simpset
+addsimp (Theorem (Prop t) _stats) ss =
+  case ruleOfProp t of
+    Nothing -> fail "addsimp: theorem not an equation"
```

Comment:
I'm not entirely settled on the right answer here, but I've been using `throwTopLevel` in some code recently when working in the `TopLevel` monad. It doesn't hugely matter, because `fail` calls `throwTopLevel` in this case, but I do slightly like the idea of avoiding having calls to `fail` in the codebase, even if we know they're well-behaved.

Reply:
I totally agree that `fail` is not the way to go in the long term. However there are already lots of other `TopLevel` primitives that use `fail` in their implementations right now, so I just used `fail` for consistency. I was thinking that we could remove all the uses of `fail` in the `TopLevel` monad in a separate PR.

Reply:
Yeah, I think that makes sense. The same question will come up shortly when I merge the external ABC interaction branch. I did modify some code to use `throwTopLevel`, but there's a bunch more refactoring to do.

Reply:
I just opened #966, which I imagine would be reasonable to address in a single PR.